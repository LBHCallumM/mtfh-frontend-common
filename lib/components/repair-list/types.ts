export type Status =
  | "AcctHold"
  | "Assigned"
  | "Canceled"
  | "Closed"
  | "Complete"
  | "Estimating"
  | "Hold"
  | "Open"
  | "VariationPendingApproval"
  | "PendDesign"
  | "PendMaterial"
  | "Scheduled"
  | "Superceded"
  | "Locked"
  | "NoAccess"
  | "PendingApproval"
  | "VariationApproved"
  | "VariationRejected";

export interface Repair {
  reference: number;
  dateRaised: string;
  lastUpdated: string | null;
  priority: string;
  property: string;
  owner: string;
  description: string;
  propertyReference: string;
  tradeCode: string;
  tradeDescription: string;
  status: string;
}
