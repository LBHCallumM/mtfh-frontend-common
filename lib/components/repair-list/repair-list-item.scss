@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/settings/colour-palette";

.repair-list-item {
  a.repair-list__link:link {
    color: lbh-colour(lbh-a01);
  }
  &__status {
    font-weight: $lbh-font-weight-bold;
  }
  &__trim {
    &::after {
      content: "...";
    }
  }
}
