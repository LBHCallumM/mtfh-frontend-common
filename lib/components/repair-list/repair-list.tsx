import React from "react";
import { config } from "../../config";
import locale from "../../locale";
import { CardList } from "../card-list";
import { Center } from "../center";
import { ErrorSummary } from "../error-summary";
import { Link } from "../link";
import { Spinner } from "../spinner";
import RepairListItem from "./repair-list-item";

export const mockRepairs = [
  {
    reference: 2267941,
    dateRaised: "2020-09-07T09:01:09Z",
    lastUpdated: null,
    priority: "ELECTRICAL PLANS",
    property: "Somerford Grove Estate 103",
    owner: "HH Electrical Planned",
    description: "Please carryout full NICEIC testing to Property (2nd QTR test 2020/21)",
    propertyReference: "00010773",
    tradeCode: "EL",
    tradeDescription: "Electrical",
    status: "Locked",
  },
  {
    reference: 2254479,
    dateRaised: "2020-08-07T20:00:55Z",
    lastUpdated: null,
    priority: "GAS SERVICING",
    property: "Somerford Grove Estate 103",
    owner: "HH Gas Breakdown",
    description: "Annual Gas Service",
    propertyReference: "00010773",
    tradeCode: "GS",
    tradeDescription: "Domestic gas: servicing",
    status: "In Progress",
  },
];

interface RepairListProps {
  propertyId: string;
}

const useRepairs = (propertyId: string) => {
  return { repairs: mockRepairs, error: null };
};

export const RepairList = ({ propertyId }: RepairListProps) => {
  const { repairs, error } = useRepairs(propertyId);
  const { components } = locale;

  if (error) {
    return (
      <ErrorSummary
        id="repair-list-error"
        title={components.repairList.errors.unableToFetchRepairs}
        description={components.repairList.errors.unableToFetchRepairsDescription}
      />
    );
  }

  if (!repairs) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  return (
    <div>
      <CardList>
        {repairs.map((repair, index) => (
          <RepairListItem key={index} repair={repair} />
        ))}
      </CardList>
      <Link
        href={`${config.repairsHubAppUrl}/properties/${propertyId}`}
        isExternal
        className="repair-list__link"
      >
        {components.repairList.seeAllRepairs}
      </Link>
    </div>
  );
};
