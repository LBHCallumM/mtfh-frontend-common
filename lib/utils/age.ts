import { differenceInYears, format, isValid, parseISO } from "date-fns";

export const isUnderEighteen = (dob: string) => {
  const isValidDate = parseISO(dob);
  if (!isValid(isValidDate)) {
    return true;
  }

  const dobDate = format(new Date(dob), "yyyy-MM-dd");
  const ageInYears = differenceInYears(new Date(), new Date(dobDate));
  return ageInYears < 18;
};
