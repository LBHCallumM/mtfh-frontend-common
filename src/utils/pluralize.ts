const pluralize = (word: string, value: number): string =>
  `${word}${Math.abs(value) !== 1 ? 's' : ''}`;

export default pluralize;
