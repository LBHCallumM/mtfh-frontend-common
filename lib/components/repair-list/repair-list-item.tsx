import React, { useMemo } from "react";

import cn from "classnames";

import { formatDate } from "@mtfh/common/lib/utils";

import { config } from "../../config";
import locale from "../../locale";

import "./repair-list-item.scss";
import { CardListBreak, CardListItem } from "../card-list";
import { CardListRows } from "../card-list/card-list-rows";
import { Link } from "../link";
import { LinkBox, LinkOverlay } from "../link-box";
import { Repair } from "./types";

export interface RepairListItemParameters {
  repair: Repair;
}

const DESCRIPTION_LENGTH = 50;

const RepairListItem = ({
  repair: { dateRaised, priority, tradeDescription, status, description, reference },
}: RepairListItemParameters): JSX.Element => {
  const { components } = locale;
  const dateRaisedAt = useMemo(() => formatDate(dateRaised), [dateRaised]);

  const rows = [
    { value: dateRaisedAt, label: components.repairList.raisedAt },
    { value: priority, label: components.repairList.priority },
  ];
  return (
    <LinkBox>
      <CardListItem>
        <LinkOverlay>
          <Link
            className="lbh-link"
            isExternal
            href={`${config.repairsHubAppUrl}/work-orders/${reference}`}
          >
            <span
              className={cn({
                "repair-list-item__trim": description.length > DESCRIPTION_LENGTH,
              })}
            >
              {tradeDescription}: {description.substring(0, DESCRIPTION_LENGTH)}
            </span>
          </Link>
        </LinkOverlay>
        <CardListRows rows={rows} />
        <CardListBreak />
        <div className="repair-list-item__status"> {status}</div>
      </CardListItem>
    </LinkBox>
  );
};

export default RepairListItem;
