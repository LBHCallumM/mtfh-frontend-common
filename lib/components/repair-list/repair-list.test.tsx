import React from "react";
import { render } from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";
import locale from "../../locale";
import { formatDate } from "../../utils";
import { RepairList } from "./repair-list";

export const mockRepairs = [
  {
    reference: 2267941,
    dateRaised: "2020-09-07T09:01:09Z",
    lastUpdated: null,
    priority: "ELECTRICAL PLANS",
    property: "Somerford Grove Estate 103",
    owner: "HH Electrical Planned",
    description: "Please carryout full NICEIC testing to Property (2nd QTR test 2020/21)",
    propertyReference: "00010773",
    tradeCode: "EL",
    tradeDescription: "Electrical",
    status: "Locked",
  },
  {
    reference: 2254479,
    dateRaised: "2020-08-07T20:00:55Z",
    lastUpdated: null,
    priority: "GAS SERVICING",
    property: "Somerford Grove Estate 103",
    owner: "HH Gas Breakdown",
    description: "Annual Gas Service",
    propertyReference: "00010773",
    tradeCode: "GS",
    tradeDescription: "Domestic gas: servicing",
    status: "In Progress",
  },
];

test("RepairList renders", () => {
  const { container } = render(<RepairList propertyId="00023400" />);

  expect(
    screen.getByText("Electrical: Please carryout full NICEIC testing to Property (2"),
  ).toBeInTheDocument();
  expect(screen.getByText(formatDate("2020-09-07T09:01:09Z"))).toBeInTheDocument();
  screen.getByText("ELECTRICAL PLANS");
  screen.getByText("Locked");

  expect(
    screen.getByText("Domestic gas: servicing: Annual Gas Service"),
  ).toBeInTheDocument();
  expect(screen.getByText(formatDate("2020-08-07T20:00:55Z"))).toBeInTheDocument();
  screen.getByText("GAS SERVICING");
  screen.getByText("In Progress");

  expect(screen.getAllByText("Raised at:").length).toBe(2);
  expect(screen.getAllByText("Priority:").length).toBe(2);

  screen.getByText(locale.components.repairList.seeAllRepairs);

  expect(container).toMatchSnapshot();
});
