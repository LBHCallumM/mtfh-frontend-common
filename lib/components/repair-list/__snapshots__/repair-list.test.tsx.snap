// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepairList renders 1`] = `
<div>
  <div>
    <div
      class="mtfh-card-list"
    >
      <div
        class="mtfh-link-box"
      >
        <div
          class="mtfh-card-list-item"
        >
          <div
            class="mtfh-link-overlay"
          >
            <a
              class="govuk-link lbh-link lbh-link"
              href="/api/v1/work-orders/2267941"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="repair-list-item__trim"
              >
                Electrical
                : 
                Please carryout full NICEIC testing to Property (2
              </span>
            </a>
          </div>
          <dl
            class="mtfh-card-list-rows"
          >
            <div
              class="mtfh-card-list-rows__row"
            >
              <dt
                class="mtfh-card-list-rows__label"
              >
                Raised at
                :
              </dt>
              <dd
                class="mtfh-card-list-rows__value"
              >
                07/09/2020
              </dd>
            </div>
            <div
              class="mtfh-card-list-rows__row"
            >
              <dt
                class="mtfh-card-list-rows__label"
              >
                Priority
                :
              </dt>
              <dd
                class="mtfh-card-list-rows__value"
              >
                ELECTRICAL PLANS
              </dd>
            </div>
          </dl>
          <hr
            class="mtfh-card-list-break"
          />
          <div
            class="repair-list-item__status"
          >
             
            Locked
          </div>
        </div>
      </div>
      <div
        class="mtfh-link-box"
      >
        <div
          class="mtfh-card-list-item"
        >
          <div
            class="mtfh-link-overlay"
          >
            <a
              class="govuk-link lbh-link lbh-link"
              href="/api/v1/work-orders/2254479"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class=""
              >
                Domestic gas: servicing
                : 
                Annual Gas Service
              </span>
            </a>
          </div>
          <dl
            class="mtfh-card-list-rows"
          >
            <div
              class="mtfh-card-list-rows__row"
            >
              <dt
                class="mtfh-card-list-rows__label"
              >
                Raised at
                :
              </dt>
              <dd
                class="mtfh-card-list-rows__value"
              >
                07/08/2020
              </dd>
            </div>
            <div
              class="mtfh-card-list-rows__row"
            >
              <dt
                class="mtfh-card-list-rows__label"
              >
                Priority
                :
              </dt>
              <dd
                class="mtfh-card-list-rows__value"
              >
                GAS SERVICING
              </dd>
            </div>
          </dl>
          <hr
            class="mtfh-card-list-break"
          />
          <div
            class="repair-list-item__status"
          >
             
            In Progress
          </div>
        </div>
      </div>
    </div>
    <a
      class="govuk-link lbh-link repair-list__link"
      href="/api/v1/properties/00023400"
      rel="noopener noreferrer"
      target="_blank"
    >
      See all live repairs in Repairs Hub &gt;
    </a>
  </div>
</div>
`;
